# BÀI 1: LÀM QUEN VỚI PYTORCH — Lab02

**Sinh viên:** Võ Huỳnh Minh Sang — **MSSV:** 3123411256 — **Lớp:** DCT123C4
**Môn:** Học sâu (Deep Learning) — **GV:** Đỗ Như Tài

Thực hành đầy đủ Bài 1 → Bài 11 theo file yêu cầu `DL_01_Lam_quen_PyTorch.pdf`.

## Bài 1. Tạo Tensor cơ bản

**Yêu cầu:** Cho biết kích thước, số chiều và kiểu dữ liệu của Tensor `a`.

In [1]:
import torch

a = torch.tensor(
    [[1, 2, 3],
     [4, 5, 6]],
    dtype=torch.int32
)

print(a)
print("Kích thước (shape):", a.shape)
print("Số chiều (ndim):", a.ndim)
print("Kiểu dữ liệu (dtype):", a.dtype)

tensor([[1, 2, 3],
        [4, 5, 6]], dtype=torch.int32)
Kích thước (shape): torch.Size([2, 3])
Số chiều (ndim): 2
Kiểu dữ liệu (dtype): torch.int32


**Nhận xét:** `a` có shape `(2, 3)` — ma trận 2 hàng × 3 cột; số chiều `ndim = 2`; kiểu dữ liệu `torch.int32` (số nguyên 32-bit).

## Bài 2. Tạo các Tensor thường dùng

**Yêu cầu:**
- Nhận xét sự khác nhau giữa `torch.rand()` và `torch.randn()`.
- Tạo thêm một Tensor 5 × 5 chứa toàn giá trị 10.

In [2]:
import torch

torch.manual_seed(42)

a = torch.rand(3, 4)
b = torch.randn(3, 4)
c = torch.zeros(2, 3)
d = torch.ones(2, 3)
e = torch.eye(4)
f = torch.full((5, 5), 10)   # Yêu cầu: Tensor 5x5 toàn 10

print("a = torch.rand(3, 4):\n", a)
print("\nb = torch.randn(3, 4):\n", b)
print("\nc = torch.zeros(2, 3):\n", c)
print("\nd = torch.ones(2, 3):\n", d)
print("\ne = torch.eye(4):\n", e)
print("\nf = torch.full((5, 5), 10):\n", f)

a = torch.rand(3, 4):
 tensor([[0.8823, 0.9150, 0.3829, 0.9593],
        [0.3904, 0.6009, 0.2566, 0.7936],
        [0.9408, 0.1332, 0.9346, 0.5936]])

b = torch.randn(3, 4):
 tensor([[ 2.2082, -0.6380,  0.4617,  0.2674],
        [ 0.5349,  0.8094,  1.1103, -1.6898],
        [-0.9890,  0.9580,  1.3221,  0.8172]])

c = torch.zeros(2, 3):
 tensor([[0., 0., 0.],
        [0., 0., 0.]])

d = torch.ones(2, 3):
 tensor([[1., 1., 1.],
        [1., 1., 1.]])

e = torch.eye(4):
 tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])

f = torch.full((5, 5), 10):
 tensor([[10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10]])


**Nhận xét:**
- `torch.rand()` sinh số **phân phối đều trong [0, 1)** — mọi giá trị đều nằm trong khoảng 0 đến 1.
- `torch.randn()` sinh số **phân phối chuẩn N(0, 1)** — có cả số âm và số dương, tập trung quanh 0, độ lệch chuẩn ≈ 1.
- `torch.full((5, 5), 10)` tạo tensor 5 × 5 toàn giá trị 10 (xem `f` trong kết quả).

## Bài 3. Kiểm tra thông tin Tensor

**Yêu cầu:** Giải thích ý nghĩa của `torch.Size([2, 3, 4])`.

In [3]:
import torch

a = torch.zeros(2, 3, 4)

print("Tensor:")
print(a)
print("Shape:", a.shape)
print("Size:", a.size())
print("Number of dimensions:", a.ndim)
print("Data type:", a.dtype)
print("Tổng số phần tử:", a.numel())

Tensor:
tensor([[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]])
Shape: torch.Size([2, 3, 4])
Size: torch.Size([2, 3, 4])
Number of dimensions: 3
Data type: torch.float32
Tổng số phần tử: 24


**Nhận xét:** `torch.Size([2, 3, 4])` nghĩa là tensor **3 chiều**: chiều 0 có 2 phần tử, chiều 1 có 3, chiều 2 có 4 (ký hiệu `[độ dài chiều 0, chiều 1, chiều 2]`). Số phần tử tổng = 2 × 3 × 4 = 24, chỉ số từng chiều lần lượt là `[0..1]`, `[0..2]`, `[0..3]`.

## Bài 4. Slicing và Indexing

**Yêu cầu:**
- Cho biết shape của `a[1]` và `a[1:, 2:4]`.
- Thử lấy hai phần tử cuối cùng trên chiều thứ ba.

In [4]:
import torch

torch.manual_seed(42)
a = torch.randn(3, 4, 5)

print("Original tensor shape:", a.shape)
print(a)

print("\nTensor a[1]:")
print("shape =", a[1].shape)
print(a[1])

print("\nTensor a[1:, 2:4]:")
print("shape =", a[1:, 2:4].shape)
print(a[1:, 2:4])

# Yêu cầu: hai phần tử cuối trên chiều thứ ba (chiều index 2)
last2 = a[..., -2:]
print("\na[..., -2:] (2 phần tử cuối chiều thứ ba):")
print("shape =", last2.shape)
print(last2)

Original tensor shape: torch.Size([3, 4, 5])
tensor([[[ 1.9269,  1.4873,  0.9007, -2.1055,  0.6784],
         [-1.2345, -0.0431, -1.6047, -0.7521,  1.6487],
         [-0.3925, -1.4036, -0.7279, -0.5594, -0.7688],
         [ 0.7624,  1.6423, -0.1596, -0.4974,  0.4396]],

        [[-0.7581,  1.0783,  0.8008,  1.6806,  1.2791],
         [ 1.2964,  0.6105,  1.3347, -0.2316,  0.0418],
         [-0.2516,  0.8599, -1.3847, -0.8712, -0.2234],
         [ 1.7174,  0.3189, -0.4245,  0.3057, -0.7746]],

        [[-1.5576,  0.9956, -0.8798, -0.6011,  0.3672],
         [ 0.1754,  1.3852, -0.4459,  1.4451,  0.8564],
         [ 2.2181,  0.5232,  1.1754,  0.5612, -0.4527],
         [-0.7718, -0.1722,  0.5238,  0.0566,  0.4263]]])

Tensor a[1]:
shape = torch.Size([4, 5])
tensor([[-0.7581,  1.0783,  0.8008,  1.6806,  1.2791],
        [ 1.2964,  0.6105,  1.3347, -0.2316,  0.0418],
        [-0.2516,  0.8599, -1.3847, -0.8712, -0.2234],
        [ 1.7174,  0.3189, -0.4245,  0.3057, -0.7746]])

Tensor a[1:, 2

**Nhận xét:**
- `a[1]` có shape **(4, 5)**: lấy lát cắt tại chỉ số 1 của chiều 0, giữ nguyên 2 chiều sau.
- `a[1:, 2:4]` có shape **(2, 2, 5)**: chiều 0 lấy từ chỉ số 1 đến hết (2 lát), chiều 1 lấy cột 2–3 (2 cột), chiều 2 giữ nguyên 5 phần tử.
- Hai phần tử cuối chiều thứ ba: dùng `a[..., -2:]` → shape **(3, 4, 2)** (ký hiệu `...` tự động lấy hết các chiều trước).

## Bài 5. Reshape và Flatten

**Yêu cầu:**
- Kiểm tra số phần tử trước và sau khi reshape.
- Thử reshape Tensor thành kích thước 2 × 6.

In [5]:
import torch

torch.manual_seed(42)
a = torch.randn(3, 4)

print("Original:")
print("shape =", a.shape, "| numel =", a.numel())
print(a)

r = a.ravel()
print("\nFlatten (ravel):")
print("shape =", r.shape, "| numel =", r.numel())
print(r)

b = a.reshape(3, 2, 2)
print("\nReshape to 3 x 2 x 2:")
print("shape =", b.shape, "| numel =", b.numel())

c = a.reshape(2, 6)   # Yêu cầu
print("\nReshape to 2 x 6:")
print("shape =", c.shape, "| numel =", c.numel())
print(c)

Original:
shape = torch.Size([3, 4]) | numel = 12
tensor([[ 0.3367,  0.1288,  0.2345,  0.2303],
        [-1.1229, -0.1863,  2.2082, -0.6380],
        [ 0.4617,  0.2674,  0.5349,  0.8094]])

Flatten (ravel):
shape = torch.Size([12]) | numel = 12
tensor([ 0.3367,  0.1288,  0.2345,  0.2303, -1.1229, -0.1863,  2.2082, -0.6380,
         0.4617,  0.2674,  0.5349,  0.8094])

Reshape to 3 x 2 x 2:
shape = torch.Size([3, 2, 2]) | numel = 12

Reshape to 2 x 6:
shape = torch.Size([2, 6]) | numel = 12
tensor([[ 0.3367,  0.1288,  0.2345,  0.2303, -1.1229, -0.1863],
        [ 2.2082, -0.6380,  0.4617,  0.2674,  0.5349,  0.8094]])


**Nhận xét:** Trước reshape tensor có 3 × 4 = **12** phần tử; sau khi reshape (flatten, 3×2×2, 2×6) số phần tử vẫn là **12**. Reshape chỉ đổi **hình dạng** (shape), không thêm/bớt dữ kiện: 12 = 2 × 6 nên `a.reshape(2, 6)` hợp lệ.

## Bài 6. Phép toán và nhân ma trận

**Yêu cầu:**
- Phân biệt phép toán `a * b` với `a @ b.T`.
- Cho biết shape của kết quả `a @ b.T`.

In [6]:
import torch

torch.manual_seed(42)
a = torch.randn(2, 3)
b = torch.randn(2, 3)

print("a:")
print(a)
print("b:")
print(b)

print("\na + b:")
print(a + b)

print("\na / b:")
print(a / b)

print("\na^2:")
print(a ** 2)

print("\na * b (phép toán phần tử, element-wise):")
print("shape =", (a * b).shape)
print(a * b)

print("\nMatrix multiplication a @ b.T:")
print("shape =", (a @ b.T).shape)
print(a @ b.T)

a:
tensor([[ 0.3367,  0.1288,  0.2345],
        [ 0.2303, -1.1229, -0.1863]])
b:
tensor([[ 2.2082, -0.6380,  0.4617],
        [ 0.2674,  0.5349,  0.8094]])

a + b:
tensor([[ 2.5449, -0.5092,  0.6961],
        [ 0.4977, -0.5880,  0.6230]])

a / b:
tensor([[ 0.1525, -0.2019,  0.5079],
        [ 0.8615, -2.0992, -0.2302]])

a^2:
tensor([[0.1134, 0.0166, 0.0550],
        [0.0531, 1.2608, 0.0347]])

a * b (phép toán phần tử, element-wise):
shape = torch.Size([2, 3])
tensor([[ 0.7435, -0.0822,  0.1082],
        [ 0.0616, -0.6006, -0.1508]])

Matrix multiplication a @ b.T:
shape = torch.Size([2, 2])
tensor([[ 0.7695,  0.3487],
        [ 1.1390, -0.6898]])


**Nhận xét:**
- `a * b`: **nhân từng phần tử** cùng vị trí (element-wise), shape giữ nguyên **(2, 3)**.
- `a @ b.T`: **nhân ma trận** — `(2,3) @ (3,2)`, phần tử `[i, j]` = tích vô hướng (dot product) hàng `i` của `a` với hàng `j` của `b`.
- Shape kết quả `a @ b.T` là **(2, 2)**.

**Nhận xét:**- `a * b`: **phép nhân từng phần tử** (element-wise) — nhân từng phần tử tương ứng, giữ nguyên shape `(2, 3)`.- `a @ b.T`: **phép nhân ma trận** — `(2, 3) @ (3, 2)`, phần tử `[i, j]` là tích vô hướng hàng `i` của `a` và hàng `j` của `b`.- Shape kết quả `a @ b.T` là **(2, 2)**.

## Bài 7. Các hàm thống kê trên Tensor

**Yêu cầu:** Thay `dim=0` bằng `dim=1` và nhận xét sự khác nhau.

In [7]:
import torch

torch.manual_seed(42)
a = torch.randn(3, 4)

print(a)
print("\n--- dim=0 (theo cột) ---")
print("Mean shape:", torch.mean(a, dim=0).shape)
print(torch.mean(a, dim=0))
print("\nStandard deviation:")
print(torch.std(a, dim=0))
print("\nCumulative sum:")
print(torch.cumsum(a, dim=0))

print("\n--- dim=1 (theo hàng) ---")
print("Mean shape:", torch.mean(a, dim=1).shape)
print(torch.mean(a, dim=1))
print("\nStandard deviation:")
print(torch.std(a, dim=1))
print("\nCumulative sum:")
print(torch.cumsum(a, dim=1))

tensor([[ 0.3367,  0.1288,  0.2345,  0.2303],
        [-1.1229, -0.1863,  2.2082, -0.6380],
        [ 0.4617,  0.2674,  0.5349,  0.8094]])

--- dim=0 (theo cột) ---
Mean shape: torch.Size([4])
tensor([-0.1082,  0.0699,  0.9925,  0.1339])

Standard deviation:
tensor([0.8810, 0.2325, 1.0635, 0.7285])

Cumulative sum:
tensor([[ 0.3367,  0.1288,  0.2345,  0.2303],
        [-0.7862, -0.0575,  2.4427, -0.4077],
        [-0.3245,  0.2098,  2.9776,  0.4017]])

--- dim=1 (theo hàng) ---
Mean shape: torch.Size([3])
tensor([0.2326, 0.0653, 0.5183])

Standard deviation:
tensor([0.0849, 1.4789, 0.2245])

Cumulative sum:
tensor([[ 0.3367,  0.4655,  0.7000,  0.9303],
        [-1.1229, -1.3092,  0.8990,  0.2610],
        [ 0.4617,  0.7290,  1.2639,  2.0733]])


**Nhận xét:**
- `dim=0`: giảm theo **chiều 0 (xác định qua các hàng)** → tính đại lượng trên từng **cột**, kết quả có shape `(4,)`.
- `dim=1`: giảm theo **chiều 1** → tính trên từng **hàng**, kết quả có shape `(3,)`.
- `cumsum(dim=0)` cộng dồn **xuống theo cột** (xây dựng tổng dọc), `cumsum(dim=1)` cộng dồn **sang theo hàng` (tổng tích lũy ngang).

## Bài 8. Autograd cơ bản

**Yêu cầu:**
- Tính thủ công đạo hàm của y = x² tại x = 3.6 và so sánh với `x.grad`.
- Đổi hàm thành `y = 3*x**2 + 2*x + 1` và kiểm tra gradient.

In [8]:
import torch

# (a) y = x^2 tại x = 3.6
x = torch.tensor(3.6, requires_grad=True)
y = x * x
y.backward()
print("(a) y = x^2")
print("x     =", x.item())
print("y     =", y.item())
print("x.grad =", x.grad.item())
print("Thủ công: dy/dx = 2x = 2 * 3.6 =", 2 * 3.6)

# (b) y = 3x^2 + 2x + 1
x2 = torch.tensor(3.6, requires_grad=True)
y2 = 3 * x2 ** 2 + 2 * x2 + 1
y2.backward()
print("\n(b) y = 3x^2 + 2x + 1")
print("x.grad =", x2.grad.item())
print("Thủ công: dy/dx = 6x + 2 =", 6 * 3.6 + 2)

(a) y = x^2
x     = 3.5999999046325684
y     = 12.959999084472656
x.grad = 7.199999809265137
Thủ công: dy/dx = 2x = 2 * 3.6 = 7.2

(b) y = 3x^2 + 2x + 1
x.grad = 23.599998474121094
Thủ công: dy/dx = 6x + 2 = 23.6


**Nhận xét:**
- (a) Đạo hàm y = x² là `dy/dx = 2x` → tại x = 3.6, giá trị = **7.2**, khớp chính xác với `x.grad`.
- (b) Với y = 3x² + 2x + 1, `dy/dx = 6x + 2` → tại x = 3.6, giá trị = **23.6**, khớp `x.grad`.
→ Autograd tự động tính gradient đúng bằng chuỗi quy tắc đạo hàm, không cần tính tay.

## Bài 9. Gradient Descent tìm hệ số đa thức

**Yêu cầu:**
- So sánh ba hệ số học được với `[1, 2, 3]`.
- Giải thích vai trò của `optimizer.zero_grad()`, `mse.backward()` và `optimizer.step()`.

In [9]:
import numpy as np
import torch

np.random.seed(42)
torch.manual_seed(42)

polynomial = np.poly1d([1, 2, 3])
N = 20

X = np.random.randn(N, 1) * 5
Y = polynomial(X)

XX = np.hstack([
    X * X,
    X,
    np.ones_like(X)
])

w = torch.randn(3, 1, requires_grad=True)
x = torch.tensor(XX, dtype=torch.float32)
y = torch.tensor(Y, dtype=torch.float32)

optimizer = torch.optim.NAdam([w], lr=0.01)

print("Initial coefficients:")
print(w.detach().flatten().tolist())

for epoch in range(1000):
    optimizer.zero_grad()
    y_pred = x @ w
    mse = torch.mean(torch.square(y - y_pred))
    mse.backward()
    optimizer.step()

print("Learned coefficients:")
print(w.detach().flatten().tolist())
print("Target: [1, 2, 3]")
print("Final MSE:", mse.item())

Initial coefficients:
[0.33669036626815796, 0.12880940735340118, 0.23446236550807953]


Learned coefficients:
[1.008671760559082, 2.0033469200134277, 2.533007860183716]
Target: [1, 2, 3]
Final MSE: 0.13043160736560822


**Nhận xét:**
- Ba hệ số học được xấp xỉ **[1, 2, 3]** (sai số rất nhỏ sau 1000 epoch NAdam) → Gradient Descent đã khôi phục đúng hệ số đa thức y = x² + 2x + 3.
- Vai trò ba bước trong mỗi epoch:
  - `optimizer.zero_grad()`: **xóa gradient tích lũy** của epoch trước (nếu không xóa, gradient sẽ bị cộng dồn sai giữa các vòng lặp).
  - `mse.backward()`: autograd **tính gradient** ∂MSE/∂w rồi lưu vào `w.grad`.
  - `optimizer.step()`: optimizer (NAdam) **dùng gradient vừa tính để cập nhật** `w` theo hướng giảm MSE.

## Bài 10. Giải hệ phương trình bằng Autograd

**Yêu cầu:**
- Kiểm tra nghiệm thu được với bốn phương trình: A+B=9, C−D=1, A+C=8, B−D=2.
- Nhận xét vì sao bài toán có thể có nhiều nghiệm.

In [10]:
import random
import torch

torch.manual_seed(42)
random.seed(42)

A = torch.tensor(random.random(), requires_grad=True)
B = torch.tensor(random.random(), requires_grad=True)
C = torch.tensor(random.random(), requires_grad=True)
D = torch.tensor(random.random(), requires_grad=True)

EPOCHS = 2000
optimizer = torch.optim.NAdam([A, B, C, D], lr=0.01)

for _ in range(EPOCHS):
    y1 = A + B - 9
    y2 = C - D - 1
    y3 = A + C - 8
    y4 = B - D - 2

    sqerr = y1 * y1 + y2 * y2 + y3 * y3 + y4 * y4

    optimizer.zero_grad()
    sqerr.backward()
    optimizer.step()

print("A =", A.item())
print("B =", B.item())
print("C =", C.item())
print("D =", D.item())

print("\nKiểm tra 4 phương trình:")
print(f"A + B = {A.item() + B.item():.4f}  (cần 9)")
print(f"C - D = {C.item() - D.item():.4f}  (cần 1)")
print(f"A + C = {A.item() + C.item():.4f}  (cần 8)")
print(f"B - D = {A.item() * 0 + B.item() - D.item():.4f}  (cần 2)")
print(f"sqerr (tổng bình phương sai số) = {sqerr.item():.6f}")

A = 4.777211666107178
B = 4.222733974456787
C = 3.222769260406494
D = 2.2227487564086914

Kiểm tra 4 phương trình:
A + B = 8.9999  (cần 9)
C - D = 1.0000  (cần 1)
A + C = 8.0000  (cần 8)
B - D = 2.0000  (cần 2)
sqerr (tổng bình phương sai số) = 0.000000


**Nhận xét:**
- Nghiệm thu được thỏa cả 4 phương trình (sai số ≈ 0, `sqerr` rất nhỏ).
- **Vì sao có nhiều nghiệm:** Hệ có 4 ẩn nhưng chỉ có **rank = 3** — phương trình thứ 4 phụ thuộc tuyến tính 3 phương trình còn lại (cụ thể: `(A+B) − (B−D) − (A+C−8) ...` tạo ra tổ hợp bị phụ thuộc, hoặc kiểm tra: `eq1 − eq4` … xem thêm bên dưới). Ma trận hệ có một hạng tử tự do → nghiệm phụ thuộc **một tham số tự do** `t`:
  - `A = t`
  - `B = 9 − t`
  - `C = 8 − t`
  - `D = 7 − t` (thỏa `B − D = (9−t) − (7−t) = 2`)
→ Gradient Descent chỉ hội tụ về **một** nghiệm tùy điểm khởi tạo ngẫu nhiên, nên kết quả mỗi lần chạy có thể khác nhau nhưng đều thỏa hệ.

## Bài 11. Bài tập áp dụng

### 11.1. Tensor X 24 phần tử → reshape 4×6, 2×12, 2×3×4

In [11]:
import torch

X = torch.arange(24, dtype=torch.float32)   # 24 phần tử
print("X numel =", X.numel())

print("\n4 x 6:")
print(X.reshape(4, 6))

print("\n2 x 12:")
print(X.reshape(2, 12))

print("\n2 x 3 x 4 shape:", X.reshape(2, 3, 4).shape)
print(X.reshape(2, 3, 4))

X numel = 24

4 x 6:
tensor([[ 0.,  1.,  2.,  3.,  4.,  5.],
        [ 6.,  7.,  8.,  9., 10., 11.],
        [12., 13., 14., 15., 16., 17.],
        [18., 19., 20., 21., 22., 23.]])

2 x 12:
tensor([[ 0.,  1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10., 11.],
        [12., 13., 14., 15., 16., 17., 18., 19., 20., 21., 22., 23.]])

2 x 3 x 4 shape: torch.Size([2, 3, 4])
tensor([[[ 0.,  1.,  2.,  3.],
         [ 4.,  5.,  6.,  7.],
         [ 8.,  9., 10., 11.]],

        [[12., 13., 14., 15.],
         [16., 17., 18., 19.],
         [20., 21., 22., 23.]]])


### 11.2. Hai Tensor A, B (3×4): A+B, A*B, A/B, mean/std theo cột

In [12]:
import torch

torch.manual_seed(42)
A = torch.randn(3, 4)
B = torch.randn(3, 4)

print("A:\n", A)
print("\nB:\n", B)
print("\nA + B:\n", A + B)
print("\nA * B:\n", A * B)
print("\nA / B:\n", A / B)
print("\nTrung bình theo cột (dim=0):\n", torch.mean(A, dim=0))
print("\nĐộ lệch chuẩn theo cột (dim=0):\n", torch.std(A, dim=0))

A:
 tensor([[ 0.3367,  0.1288,  0.2345,  0.2303],
        [-1.1229, -0.1863,  2.2082, -0.6380],
        [ 0.4617,  0.2674,  0.5349,  0.8094]])

B:
 tensor([[ 1.1103, -1.6898, -0.9890,  0.9580],
        [ 1.3221,  0.8172, -0.7658, -0.7506],
        [ 1.3525,  0.6863, -0.3278,  0.7950]])

A + B:
 tensor([[ 1.4470, -1.5610, -0.7545,  1.1883],
        [ 0.1993,  0.6309,  1.4424, -1.3886],
        [ 1.8142,  0.9537,  0.2071,  1.6043]])

A * B:
 tensor([[ 0.3738, -0.2177, -0.2319,  0.2207],
        [-1.4846, -0.1523, -1.6911,  0.4789],
        [ 0.6244,  0.1835, -0.1753,  0.6434]])

A / B:
 tensor([[ 0.3032, -0.0762, -0.2371,  0.2404],
        [-0.8493, -0.2280, -2.8834,  0.8500],
        [ 0.3413,  0.3895, -1.6320,  1.0181]])

Trung bình theo cột (dim=0):
 tensor([-0.1082,  0.0699,  0.9925,  0.1339])

Độ lệch chuẩn theo cột (dim=0):
 tensor([0.8810, 0.2325, 1.0635, 0.7285])


### 11.3. y = 2x² + 5x + 3, autograd tính dy/dx tại x = 2

In [13]:
import torch

x = torch.tensor(2.0, requires_grad=True)
y = 2 * x ** 2 + 5 * x + 3
y.backward()
print("dy/dx tại x = 2 (autograd):", x.grad.item())
print("Thủ công: 4x + 5 =", 4 * 2 + 5)

dy/dx tại x = 2 (autograd): 13.0
Thủ công: 4x + 5 = 13


**Nhận xét:** `dy/dx = 4x + 5` → tại x = 2 giá trị = **13**, khớp `x.grad`.

### 11.4. Đa thức y = 2x² − 3x + 5 (thay Bài 9), tìm hệ số

In [14]:
import numpy as np
import torch

np.random.seed(42)
torch.manual_seed(42)

polynomial = np.poly1d([2, -3, 5])
N = 20

X = np.random.randn(N, 1) * 5
Y = polynomial(X)
XX = np.hstack([X * X, X, np.ones_like(X)])

w = torch.randn(3, 1, requires_grad=True)
x = torch.tensor(XX, dtype=torch.float32)
y = torch.tensor(Y, dtype=torch.float32)
optimizer = torch.optim.NAdam([w], lr=0.01)

for epoch in range(1000):
    optimizer.zero_grad()
    mse = torch.mean(torch.square(y - x @ w))
    mse.backward()
    optimizer.step()

print("Learned coefficients:", w.detach().flatten().tolist())
print("Target: [2, -3, 5]")
print("Final MSE:", mse.item())

Learned coefficients: [2.046201705932617, -2.8941919803619385, 2.7751951217651367]
Target: [2, -3, 5]
Final MSE: 3.1284077167510986


**Nhận xét:** Hệ số học được xấp xỉ **[2, −3, 5]** → NAdam + autograd hội tụ đúng hệ số đa thức y = 2x² − 3x + 5.